spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/logging/PassthroughLogger.java

```diff
@@ -0,0 +1,80 @@
+package org.springframework.cloud.gateway.logging;
+
+import org.apache.commons.logging.Log;
+import org.springframework.web.server.ServerWebExchange;
+
+public class PassthroughLogger implements AdaptableLogger {
```

Comment:
Maybe there is another interface `AdaptableLogger.Factory` and that's the thing passed into the class. And then you create one of these things passing `getClass()` as the constructor and then the `Log` is just a field in this class.

Reply:
We could also introduce `Supplier` variants of logging methods similar to `LogAccessor` from framework.

Reply:
We could use an ObjectFactory<AdaptableLogger> that can be used to create the AdaptableLogger with the underlying log instance. This way we wouldn't have to pass the log object. This would change the invocation from:

```Logger.trace(log, exchange,"message)``` 

to 

```logger.trace(exchange, "message")```

I will update PR to illustrate.

Reply:
how would this work with teams using slf4j logging with varargs?  For example:

```
    loger.error("foo={}, bar={}", foo, bar, exception);
```

Reply:
The interface was designed with JCL in mind since this is what spring and Spring cloud gateway used. JCL doesn't have a varargs option. If you have written a customer filter / predicate and are using SLFJ then you will have to manage that log statement yourself (WebFluxSleuthOperators.withSpanInScope()). Note, you can change the Sleuth instrumentation mode to be decorate-on-last and you will get the previous Huxton behavior (but this mode has issues which is why the manual mode was created).

Reply:
while that's true, I have to concur with your comment in the issue you created:
> But it means we need to rewrite all the logging statements and refactor it to pass the exchange. It's really a big change.

I'm just hoping for some middle ground somewhere.

Reply:
Yeah, I think this is an unusual situation that, until the `decorate-on-last` situation is performant, we'll have to live with it.

Reply:
Do you have any stats on the performance issue with using `decorate-on-each`?  Our own tests didn't show any significant issue in previous releases.  We just can't do this massive of a rewrite and we will have to rely on this instrumentation-type, and I'm looking for something to show the folks around here.

Reply:
Yep, see [here](https://github.com/tony-clarke-amdocs/sleuth-reactive-perf/tree/2020.0.0). It's pretty bad.

Reply:
BTW, it's not just performance. decorate-on-each and decorate-on-last can both drop the trace id in certain circumstances. The only method that will work in all situations is manual. Hence, this PR.

Reply:
my concern is that this PR only addresses one specific situation, and that's logging in Gateway-specific filters.  There's a much larger problem with custom filters, non-Gateway code, not to mention just OpenTracing in general (OpenTracing's Tracer.activeSpan() no longer works w/o decorate-on-each).  I'm hoping that there is a long-term path where we can rely on decorate-on-each to cover these normal situations where there's no access to a ServerWebExchange.

Reply:
As far as I know there is no plan to improve decorate-on-each or decorate-on-last. @marcingrzejszczak will correct me if I have it wrong.

Reply:
Actually there is a PR that seems to be improving the performance of the decorate on each option https://github.com/spring-cloud/spring-cloud-sleuth/pull/1830